mypy/types.py

```diff
@@ -144,6 +144,97 @@ def deserialize(cls, data: JsonDict) -> 'Type':
         raise NotImplementedError('Cannot deserialize {} instance'.format(cls.__name__))
 
 
+class TypeAliasType(Type):
+    """
+    A type alias to another type.
+
+    To support recursive type aliases we don't immediately expand a type alias
+    during semantic analysis, but create an in stance of this type that records the target alias
+    definition node (mypy.nodes.TypeAlias) and type arguments (for generic aliases).
+
+    This is very similar to how TypeInfo vs Instance interact.
+    """
+    def __init__(self, alias: Optional[mypy.nodes.TypeAlias], args: List[Type],
+                 line: int = -1, column: int = -1) -> None:
+        super().__init__(line, column)
+        self.alias = alias
+        self.args = args
+        self.type_ref = None  # type: Optional[str]
+
+    def expand_once(self) -> Type:
+        """Expand to the target type exactly once.
+
+        This doesn't do full expansion, i.e. the result can contain another
+        (or even this same) type alias.
```

Comment:
Maybe this should be considered an internal helper (perhaps prefixed with `_`) and we could suggest using `get_proper_type` instead.